kolibri/plugins/learn/assets/src/actions.js

```diff
@@ -1,76 +1,88 @@
-    });
-  }
-};
+const PageModes = constants.PageModes;
+const PageNames = constants.PageNames;
 
-/**
- * Function to dispatch mutations to topics and contents by node kind.
- * @param {Object[]} nodes - Data to dispatch mutations with.
- * @param {Function} dispatch - dispatch method of Vuex Store.
- */
-const nodeAssignment = (nodes, dispatch) => {
-  const topics = nodes.filter((node) => node.kind === 'topic');
-  const contents = nodes.filter((node) => node.kind !== 'topic');
 
-  // clean up API response
-  contents.forEach(content => {
-    if (!content.progress) {
-      content.progress = 'unstarted'; // eslint-disable-line no-param-reassign
-    }
-    if (!content.thumbnail) {
-      content.files.forEach(file => {
-        if (file.thumbnail) {
-          content.thumbnail = file.storage_url; // eslint-disable-line no-param-reassign
-        }
-      });
+function _crumbState(ancestors) {
+  return ancestors.map((ancestor, index) => {
+    if (index === 0) {
+      return { id: 'root', title: 'Home' }; // TODO - I18N
     }
+    return {
+      id: ancestor.pk,
+      title: ancestor.title,
+    };
   });
+}
 
-  dispatch('SET_TOPICS', topics);
-  dispatch('SET_CONTENTS', contents);
-};
+function _topicState(data, includeCrumbs = true) {
+  const state = {
+    id: data.pk,
+    title: data.title,
+    description: data.description,
+  };
+  if (includeCrumbs) {
+    state.breadcrumbs = _crumbState(data.ancestors);
+  }
+  return state;
+}
 
-/**
- * Action to fetch child topics of a particular topic from the API.
- * @param {Function} dispatch - The dispatch method of the store object.
- * @param {String} id - The id of the model to fetch the children of.
- */
-const fetchNodes = ({ dispatch }, id) => {
-  // Get the collection from ContentNodeResource.
-  if (typeof id === 'undefined') {
-    id = 'root'; // eslint-disable-line no-param-reassign
+function _contentState(data, includeCrumbs = true) {
+  const state = {
+    id: data.pk,
+    title: data.title,
+    description: data.description,
+    thumbnail: data.thumbnail,
+    files: data.files,
+    progress: data.description ? data.description : 'unstarted',
+  };
+  if (includeCrumbs) {
+    state.breadcrumbs = _crumbState(data.ancestors);
   }
-  const contentCollection = Kolibri.resources.ContentNodeResource.getCollection({ parent: id });
-  if (contentCollection.synced) {
-    nodeAssignment(contentCollection.data, dispatch);
-  } else {
-    contentCollection.fetch().then(() => {
-      nodeAssignment(contentCollection.data, dispatch);
+  return state;
+}
+
+function navToExploreTopic(store, id) {
+  store.dispatch('SET_PAGE_NAME', PageNames.EXPLORE_ROOT);
+  store.dispatch('SET_PAGE_MODE', PageModes.EXPLORE);
+  store.dispatch('SET_LOADING');
+
+  const pageState = { id };
+
+  Resources.getModel(id).fetch()
+    .then((topicAttributes) => {
+      pageState.topic = _topicState(topicAttributes);
+      return Resources.getCollection({ parent: id }).fetch();
```

Comment:
Given that we have the 'id' already - not sure why these have to be chained. Could do both asynchronously.